src/util.rs

```diff
@@ -1,23 +1,91 @@
+            // Several possible options and their resolutions:
+            // (1) The file isn't actually a rustdoc JSON file. The user should supply a valid file.
+            // (2) The rustdoc JSON file has a version number that is too old, and isn't supported.
+            //     The user should upgrade to a newer nightly Rust version and regenerate the file.
+            // (3) The rustdoc JSON file has a version number that is too new, and isn't supported.
+            //     The user should attempt to upgrade to a newer cargo-semver-checks version
+            //     if one is already available, or open a GitHub issue otherwise.
+
+            // The error on this line is case (1).
+            let version = serde_json::from_str::<RustdocFormatVersion>(&s).with_context(|| {
+                format!(
+                    "parsing rustdoc JSON failed, and could not determine rustdoc format version \
+                    for file {}",
+                    path.display(),
+                )
+            })?;
 
-    serde_json::from_str(&s)
-        .with_context(|| format!("Failed to parse rustdoc JSON output file {:?}", path))
+            match version.format_version.cmp(&rustdoc_types::FORMAT_VERSION) {
+                std::cmp::Ordering::Less => {
+                    // The error here is case (2).
+                    bail!(
+                        "rustdoc output format is too old (v{1}, need v{0}) for file \
+                        {2}\n\n\
+                        \
+                        note: Using a newer Rust nightly version should help. \
+                        The latest cargo-semver-checks and Rust nightly should always be \
+                        compatible with each other.",
+                        rustdoc_types::FORMAT_VERSION,
+                        version.format_version,
+                        path.display(),
+                    )
+                }
+                std::cmp::Ordering::Greater => {
+                    // The error here is case (3).
+                    bail!(
+                        "rustdoc output format is too new (v{1}, need v{0}) for this \
+                        cargo-semver-checks when parsing {2}\n\n\
+                        \
+                        note: A newer version of cargo-semver-checks is likely available. \
+                        The latest cargo-semver-checks and Rust nightly should always be \
+                        compatible with each other.",
+                        rustdoc_types::FORMAT_VERSION,
+                        version.format_version,
+                        path.display(),
+                    )
+                }
+                std::cmp::Ordering::Equal => Err(e).with_context(|| {
+                    format!(
+                        "parsing rustdoc failed even though the format was the expected \
+                            version (v{}) for file {}\n\n\
```

Comment:
Nit: To put this more in user terms what about `Unexpected parse error for v{} rustdoc file: {}`?